feat(sitemap): add sitemap

LinHuiqing opened the pull request:
## Problem

Search crawlers may not be able to understand the structure of the site, resulting more time spent crawling unnecessary URLs on the site. This waste of time translates to a waste of crawl budget, meaning that search crawlers might not be able to index the whole site in time.

A sitemap helps in our case as our agency pages are in silos.

Closes #400 

## Solution

Use `express-sitemap-xml` to generate and serve sitemap. Sitemap is served at `/sitemap.xml`. Sitemap checkers (e.g. https://seositecheckup.com/tools/sitemap-test) should be able to see sitemap.

**Features**:

- add all paths which members of public should be able to see to the sitemap

## Before & After Screenshots

**BEFORE**:

When we `GET` `/sitemap.xml`, users will be directed to `/not-found`.

Sitemap checkers are unable to detect the sitemap either:

<img width="934" alt="Screenshot 2021-10-13 at 1 58 10 PM" src="https://user-images.githubusercontent.com/37061143/137075503-41dc5f29-1e4c-446e-9515-f78bd8ec34ad.png">


**AFTER**:

Sitemap file is served at `/sitemap.xml`.

<img width="752" alt="Screenshot 2021-10-13 at 1 55 05 PM" src="https://user-images.githubusercontent.com/37061143/137075152-ccc02db3-54a4-44da-a846-abcbdf33fc2e.png">

Sitemap checker is able to detect the sitemap.

<img width="938" alt="Screenshot 2021-10-13 at 1 54 34 PM" src="https://user-images.githubusercontent.com/37061143/137075098-6a8faa85-9964-451c-9b08-8dd0066fce57.png">

## Tests

1. Check `/sitemap.xml` to ensure that a sitemap is served and that it includes static paths (visible to public), post pages and agency pages.
2. Use a sitemap checker to check that sitemap can be detected.

LinHuiqing committed:
feat(sitemap): add sitemap with static paths visible to MOPs

- (package.json) add express-sitemap-xml
- (package.json) add @types/express-sitemap-xml to dev
- serve sitemap on express
- add sitemap to robots.txt
- add host url to convict

re #400

LinHuiqing committed:
feat(sitemap): add posts dynamically to sitemap

- use postService's listPosts to get all post ids
- make getSitemapUrls async

re #400

LinHuiqing committed:
feat(sitemap): add agency pages dynamically to sitemap

- add listAgencyShortnames to agency service
- add listAgencyShortnames to mocked agency service

re #400

LoneRifle commented on server/package.json:
```diff
@@ -32,6 +32,7 @@
     "datadog-winston": "^1.5.1",
     "express": "^4.17.1",
     "express-session": "^1.17.2",
+    "express-sitemap-xml": "^3.0.0",
```
File the dep change as a separate PR so that you don't have to have a headache with merge conflicts

LinHuiqing replied:
Okays, filed PR #571.

LoneRifle commented on server/src/bootstrap/index.ts:
```diff
@@ -194,12 +196,45 @@ if (baseConfig.nodeEnv === Environment.Prod) {
 
 app.use('/api/v1', api(apiOptions))
 
+const getSitemapUrls = async () => {
```
I would implement this into its own module..

LinHuiqing replied:
okay done 👍

LoneRifle reviewed (changes requested):
Don't forget your tests =)

LoneRifle commented:
One other thing - `PostService.listPosts()` will be relatively expensive to invoke. Can you figure a way to minimise the number of invocations? If you are going to approach this with caching, you may need to devise a strategy for cache invalidation , and if so, you need to ensure that the db queries (if any!) you make to ensure sitemap freshness is cheap to run.

LinHuiqing commented:
> One other thing - `PostService.listPosts()` will be relatively expensive to invoke. Can you figure a way to minimise the number of invocations? If you are going to approach this with caching, you may need to devise a strategy for cache invalidation , and if so, you need to ensure that the db queries (if any!) you make to ensure sitemap freshness is cheap to run.

Thanks for bringing this up, I thought about these considerations but eventually left it as the current implementation due to the following reasons: 

1. Based on the [`express-sitemap-xml`](https://github.com/feross/express-sitemap-xml) documentation: 

    > The `getUrls` function is called at most once per 24 hours. The resulting sitemap(s) are cached to make repeated HTTP requests faster.
    
    This means that though the `PostService.listPosts()` function is expensive, it will only be called once per day when the server is up and running.

2. Google's search crawlers only use sitemaps as a recommendation according to [Google Search Central's documentation](https://developers.google.com/search/docs/advanced/sitemaps/overview):

    > Using a sitemap doesn't guarantee that all the items in your sitemap will be crawled and indexed, as Google processes rely on complex algorithms to schedule crawling. However, in most cases, your site will benefit from having a sitemap, and you'll never be penalized for having one.

    So to my knowledge, it would not be that detrimental if our site is slightly outdated as Google will still be able to crawl the new/ updated pages as long as they are linked somewhere in our site.

Considering the above, do you think we should still try to minimise the number of invocations and ensure sitemap freshness?...

LinHuiqing committed:
refactor(sitemap): extract getSitemapUrls into web module

- add sitemap path into web module
- documentation of various functions used
- removed additional curly braces in web controller
- removed unnecessary import in file controller

re #400

LoneRifle commented:
Good points! I would imagine that there are other crawlers out there, but that should be mostly handled by the package, as you've pointed out.

LoneRifle reviewed (approved):
lgtm. Do provide unit tests for the module if you can spare the time.
